libp2p/transports/tortransport.nim

```diff
@@ -0,0 +1,166 @@
+    trace "Failed to create observedAddr", exc = exc.msg
+    if not(isNil(client) and client.closed):
+      await client.closeWait()
+    raise exc
+
+  trace "Handling tcp connection", address = $observedAddr,
+                                   dir = $dir,
+                                   clients = self.tcpTransport.clients[Direction.In].len +
+                                   self.tcpTransport.clients[Direction.Out].len
+
+  let conn = Connection(
+    ChronosStream.init(
+      client = client,
+      dir = dir,
+      observedAddr = observedAddr
+    ))
+
+  proc onClose() {.async.} =
+    try:
+      let futs = @[client.join(), conn.join()]
+      await futs[0] or futs[1]
+      for f in futs:
+        if not f.finished: await f.cancelAndWait() # cancel outstanding join()
+
+      trace "Cleaning up client", addrs = $client.remoteAddress,
+                                  conn
+
+      self.tcpTransport.clients[dir].keepItIf( it != client )
+      await allFuturesThrowing(
+        conn.close(), client.closeWait())
+
+      trace "Cleaned up client", addrs = $client.remoteAddress,
+                                 conn
+
+    except CatchableError as exc:
+      let useExc {.used.} = exc
+      debug "Error cleaning up client", errMsg = exc.msg, conn
+
+  self.tcpTransport.clients[dir].add(client)
+  asyncSpawn onClose()
+
+  return conn
+
+proc connectToTorServer(transportAddress: TransportAddress): Future[StreamTransport] {.async, gcsafe.} =
+  let transp = await connect(transportAddress)
+  try:
+    var bytesWritten = await transp.write(@[05'u8, 01, 00])
+    var resp = await transp.read(2)
+    return transp
+  except CatchableError as err:
+    await transp.closeWait()
+    raise err
+
+proc dialPeer(transp: StreamTransport, address: MultiAddress) {.async, gcsafe.} =
+
+  let addressArray = ($address).split('/')
+  let addressStr = addressArray[2].split(':')[0] & ".onion"
+
+  # The address field contains a fully-qualified domain name.
+  # The first octet of the address field contains the number of octets of name that
+  # follow, there is no terminating NUL octet.
+  let dstAddr = @(uint8(addressStr.len).toBytes()) & addressStr.toBytes()
+  let dstPort = address.data.buffer[37..38]
+  let b = @[05'u8, 01, 00, 03] & dstAddr & dstPort
+
+  discard await transp.write(b)
+  discard await transp.read(10)
+
+method dial*(
+  self: TorTransport,
+  hostname: string,
+  address: MultiAddress): Future[Connection] {.async, gcsafe.} =
+  ## dial a peer
+  ##
+
+  trace "Dialing remote peer", address = $address
+  let transp = await connectToTorServer(self.transportAddress)
+
+  try:
+    await dialPeer(transp, address)
+    return await self.connHandler(transp, Direction.Out)
+  except CatchableError as err:
+    await transp.closeWait()
+    raise err
+
+method start*(
+  self: TorTransport,
+  addrs: seq[MultiAddress]) {.async.} =
+  ## listen on the transport
+  ##
+
+  #await procCall Transport(self).start(addrs)
```

Comment:
We still need to call this to set `running = true` (and maybe more stuff in the future)

Reply:
tcptransport start method is going to call it, isn't it? Should `running = true` be set only if the seq of handled addresses isn't empty?

Reply:
self.tcpTransport.running will be true, but not self.running

Reply:
Doesnt't tcptransport start call` procCall Transport(self).start(addrs)`?

Reply:
Oh right, forgot I'm using composition here. How about the second question?